File: app/src/lib/retry-utils.js
/**
 * Retry utility for handling rate limit errors (429) with exponential backoff
 */

const DEFAULT_MAX_RETRIES = 3;
const DEFAULT_INITIAL_DELAY = 1000; // milliseconds
const DEFAULT_MAX_DELAY = 60000; // milliseconds
const DEFAULT_BACKOFF_MULTIPLIER = 2.0;

/**
 * Check if an error is a rate limit error (429)
 * @param {Error|Object} error - The error to check
 * @returns {boolean} - True if it's a rate limit error
 */
function isRateLimitError(error) {
  // Check for axios error with 429 status
  if (error?.response?.status === 429) {
    return true;
  }
  
  // Check for HTTPException with 429 status
  if (error?.status === 429 || error?.statusCode === 429) {
    return true;
  }
  
  // Check error message for rate limit indicators
  const errorMessage = error?.message?.toLowerCase() || String(error).toLowerCase();
  if (errorMessage.includes('429') || 
      errorMessage.includes('rate limit') || 
      errorMessage.includes('quota') ||
      errorMessage.includes('too many requests')) {
    return true;
  }
  
  return false;
}

/**
 * Sleep for a given number of milliseconds
 * @param {number} ms - Milliseconds to sleep
 * @returns {Promise<void>}
 */
function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Retry a function with exponential backoff for rate limit errors
 * @param {Function} fn - The async function to retry
 * @param {Object} options - Retry options
 * @param {number} options.maxRetries - Maximum number of retries (default: 3)
 * @param {number} options.initialDelay - Initial delay in ms (default: 1000)
 * @param {number} options.maxDelay - Maximum delay in ms (default: 60000)
 * @param {number} options.backoffMultiplier - Backoff multiplier (default: 2.0)
 * @param {Function} options.retryOn - Function to determine if error should be retried (default: isRateLimitError)
 * @returns {Promise<any>} - The result of the function
 */
export async function retryWithBackoff(
  fn,
  {
    maxRetries = DEFAULT_MAX_RETRIES,
    initialDelay = DEFAULT_INITIAL_DELAY,
    maxDelay = DEFAULT_MAX_DELAY,
    backoffMultiplier = DEFAULT_BACKOFF_MULTIPLIER,
    retryOn = isRateLimitError,
  } = {}
) {
  let lastError = null;
  let delay = initialDelay;
  
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error;
      
      // Check if we should retry this error
      if (!retryOn(error)) {
        console.debug(`Error is not retryable: ${error.message || error}`);
        throw error;
      }
      
      // If this was the last attempt, throw the error
      if (attempt >= maxRetries) {
        console.error(
          `Max retries (${maxRetries}) exceeded. Last error: ${error.message || error}`
        );
        throw error;
      }
      
      // Log retry attempt
      const statusCode = error?.response?.status || error?.status || 'unknown';
      console.warn(
        `Retryable error detected (attempt ${attempt + 1}/${maxRetries + 1}). ` +
        `Status: ${statusCode}. Retrying in ${(delay / 1000).toFixed(2)} seconds... ` +
        `Error: ${error.message || error}`
      );
      
      // Wait before retrying
      await sleep(delay);
      
      // Calculate next delay with exponential backoff
      delay = Math.min(delay * backoffMultiplier, maxDelay);
    }
  }
  
  // Should never reach here, but just in case
  if (lastError) {
    throw lastError;
  }
}

/**
 * Wrapper for axios requests with automatic retry on 429 errors
 * @param {Function} axiosCall - Function that returns an axios promise
 * @param {Object} retryOptions - Retry options (same as retryWithBackoff)
 * @returns {Promise<any>} - The axios response
 */
export async function axiosWithRetry(axiosCall, retryOptions = {}) {
  return retryWithBackoff(axiosCall, retryOptions);
}

File: documents-api/utils/retry_utils.py
import logging
import time
import asyncio
from typing import Callable, Any, Optional
from functools import wraps

LOGGER = logging.getLogger(__name__)

# Default retry configuration
DEFAULT_MAX_RETRIES = 3
DEFAULT_INITIAL_DELAY = 1.0  # seconds
DEFAULT_MAX_DELAY = 60.0  # seconds
DEFAULT_BACKOFF_MULTIPLIER = 2.0


def is_rate_limit_error(exception: Exception) -> bool:
    """
    Check if an exception is a rate limit error (429).
    Handles both HTTPException and Google API errors.
    """
    # Check for HTTPException with 429 status
    if hasattr(exception, 'status_code') and exception.status_code == 429:
        return True
    
    # Check for Google API rate limit errors
    error_str = str(exception).lower()
    if '429' in error_str or 'rate limit' in error_str or 'quota' in error_str:
        return True
    
    # Check for specific Google API error types
    
    # Check for Google API client errors
    if hasattr(exception, 'code') and exception.code == 429:
        return True
    
    return False


def retry_with_backoff(
    max_retries: int = DEFAULT_MAX_RETRIES,
    initial_delay: float = DEFAULT_INITIAL_DELAY,
    max_delay: float = DEFAULT_MAX_DELAY,
    backoff_multiplier: float = DEFAULT_BACKOFF_MULTIPLIER,
    retry_on: Optional[Callable[[Exception], bool]] = None,
):
    """
    Decorator for retrying functions with exponential backoff.
    
    Args:
        max_retries: Maximum number of retry attempts (default: 3)
        initial_delay: Initial delay in seconds before first retry (default: 1.0)
        max_delay: Maximum delay in seconds between retries (default: 60.0)
        backoff_multiplier: Multiplier for exponential backoff (default: 2.0)
        retry_on: Optional function to determine if an exception should be retried.
                  If None, defaults to checking for rate limit errors (429).
    
    Usage:
        @retry_with_backoff(max_retries=5, initial_delay=2.0)
        def my_function():
            # function code
    """
    if retry_on is None:
        retry_on = is_rate_limit_error
    
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        def sync_wrapper(*args, **kwargs) -> Any:
            last_exception = None
            delay = initial_delay
            
            for attempt in range(max_retries + 1):
                try:
                    return func(*args, **kwargs)
                except Exception as e:
                    last_exception = e
                    
                    # Check if we should retry this error
                    if not retry_on(e):
                        LOGGER.debug(f"Error is not retryable: {e}")
                        raise
                    
                    # If this was the last attempt, raise the exception
                    if attempt >= max_retries:
                        LOGGER.error(
                            f"Max retries ({max_retries}) exceeded for {func.__name__}. "
                            f"Last error: {e}"
                        )
                        raise
                    
                    # Log retry attempt
                    LOGGER.warning(
                        f"Retryable error ({type(e).__name__}) in {func.__name__} (attempt {attempt + 1}/{max_retries + 1}). "
                        f"Retrying in {delay:.2f} seconds... Error: {e}"
                    )
                    
                    # Wait before retrying
                    time.sleep(delay)
                    
                    # Calculate next delay with exponential backoff
                    delay = min(delay * backoff_multiplier, max_delay)
            
            # Should never reach here, but just in case
            if last_exception:
                raise last_exception
            # Explicit return None if somehow we reach here without an exception
            return None
        
        @wraps(func)
        async def async_wrapper(*args, **kwargs) -> Any:
            last_exception = None
            delay = initial_delay
            
            for attempt in range(max_retries + 1):
                try:
                    return await func(*args, **kwargs)
                except Exception as e:
                    last_exception = e
                    
                    # Check if we should retry this error
                    if not retry_on(e):
                        LOGGER.debug(f"Error is not retryable: {e}")
                        raise
                    
                    # If this was the last attempt, raise the exception
                    if attempt >= max_retries:
                        LOGGER.error(
                            f"Max retries ({max_retries}) exceeded for {func.__name__}. "
                            f"Last error: {e}"
                        )
                        raise
                    
                    # Log retry attempt
                    LOGGER.warning(
                        f"Retryable error ({type(e).__name__}) in {func.__name__} (attempt {attempt + 1}/{max_retries + 1}). "
                        f"Retrying in {delay:.2f} seconds... Error: {e}"
                    )
                    
                    # Wait before retrying
                    await asyncio.sleep(delay)
                    
                    # Calculate next delay with exponential backoff
                    delay = min(delay * backoff_multiplier, max_delay)
            
            # Should never reach here, but just in case
            if last_exception:
                raise last_exception
            # Explicit return None if somehow we reach here without an exception
            return None
        
        # Return appropriate wrapper based on whether function is async
        import inspect
        if inspect.iscoroutinefunction(func):
            return async_wrapper
        else:
            return sync_wrapper
    
    return decorator

